File: Algorithms/basic_algorithm/SortAlgorithmsbyJava/HeapSort.java
package basic_algorithm;

public class HeapSort {

	// 用 Java 实现堆排序
	public static void heapSort(int[] arr) {
		if (arr == null || arr.length < 2) {
			return;
		}
		for (int i = 0; i < arr.length; i++) {
			heapInsert(arr, i); // 在 0~i 直接形成大根堆
		}
		// for循环 后整个数组形成大根堆结构		
		int heapSize = arr.length;
		swap(arr, 0, --heapSize);
		while (heapSize > 0) {
			heapify(arr, 0, heapSize);
			swap(arr, 0, --heapSize);			
		}
		
	}
	
	// 实现大根堆的结构，与父节点进行对比和交换	
	public static void heapInsert(int[] arr, int index) {
		while (arr[index] > arr[(index - 1)/2]) {
			swap(arr, index, (index - 1) / 2);
			index = (index - 1) / 2;			
		}
	}
	
	// index 位置被更换后如何重新变成大根堆	
	public static void heapify(int[] arr, int index, int heapSize) {
		int left = index * 2 + 1;
		while (left < heapSize) {
			// 取到左右孩子中的最大的那一个			
			int largest = left + 1 < heapSize && arr[left + 1] > arr[left] ? left+1 : left;  // 右孩子不越界，并且右孩子比左孩子大才能取到左孩子
			largest = arr[largest] > arr[index] ? largest : index;
			if (largest == index) {
				break;
			}
			swap(arr, index, largest);
			index = largest;
			left = index * 2 + 1;
		}
		
	}
	
	public static void swap(int[] arr, int i, int j) {
		// 交换两个数
		int tmp = arr[i];
		arr[i] = arr[j];
		arr[j] = tmp;
	}
}

File: Algorithms/basic_algorithm/SortAlgorithmsbyJavaScript/HeapSort.js
/**
 * @author SkylineBin
 * @time 2019-03-20
 * @function create HeapSort by JavaScript
 *  实现堆排序
 * @Algorithm_bigO 
 *  O(N * log(N))
 * 
 */

module.exports = function HeapSort(arrays) {
    if (arrays == null || arrays.length < 2) {
        return;
    }
    // 处理 快排 的 过程
    for (let index = 0; index < arrays.length; index++) {
        heapInsert(arrays, index);
    }
    let heapSize = arrays.length;
    swap(arrays, 0, --heapSize);
    while (heapSize > 0) {
        heapify(arrays, 0, heapSize);
        swap(arrays, 0, --heapSize);
    }
    
}

/**
 * @function 将一个数组调整成大根堆结构
 * @param {*} arrays 
 * @param {*} index 
 */
function heapInsert(arrays, index) {
    // 要使用 parseInt,要不然会出错
    while (arrays[index] > arrays[parseInt((index-1)/2)]) {
        swap(arrays, index, parseInt((index - 1) / 2));
        index = parseInt((index - 1) / 2);
    }
}

/**
 * @function  index位置的数变小后进行往下沉处理，直到形成新的大根堆
 * @param {*} arrays 
 * @param {*} index 
 * @param {*} heapSize 
 */
function heapify(arrays, index, heapSize){
    let left = index * 2 + 1;
    while (left < heapSize) {
        let largest = ((left + 1) < heapSize) && arrays[left+1] > arrays[left]  ? left + 1 : left;
        largest = arrays[largest] > arrays[index] ? largest : index;
        if (largest === index) {
            break;
        }
        swap(arrays, index, largest);
        index = largest;
        left = index * 2 + 1;
    }
}

/**
 * @function 用于交互数组中的指定位置的两个数
 * @param {*} arrays 
 * @param {*} si 
 * @param {*} sj 
 */
function swap(arrays, si, sj) {
    let tempdata = arrays[si];
    arrays[si] = arrays[sj];
    arrays[sj] = tempdata;
}